### ISSUE TITLE
`\numlist`, `\numproduct`, `\numrange` → `\num`

### ISSUE THREAD

**Issue by dbitouze** [josephwright/siunitx#648]
It would be handy for all the `\num...` commands (`\numlist`, `\numproduct`, `\numrange`) to just be the `\num` command, the distinction being made by the presence or absence of :

- `;` in the argument,
- `x` in the argument,
- a second argument.

Hence :

- `\num{0.1;0.2;0.3}` would behave as  `\numlist{0.1;0.2;0.3}`,
- `\num{5 x 100 x 2}` would behave as  `\numproduct{5 x 100 x 2}`,
- `\num{5}{100}` would behave as  `\numrange{5}{100}`.

Obviously, the analogous would be nice for the `\qty...` commands (`\qtylist`, `\qtyproduct`, `\qtyrange`).

**Comment by josephwright** [comment]
I split these quite deliberately as they have different semantics.

**Comment by josephwright** [comment]
There is of course a programmers API; you can create your own versions of interface commands using the documented API plus whatever 'glue' you need to achieve that (`xparse` and a `g`-type argument would cover `\numrange`, a test for a `x` plus the existing code for lists and products could then be used for those cases).

**Comment by josephwright** [comment]
Even without using the API, you could simply overload, for example
```latex
\documentclass{article}
\usepackage{siunitx}
\ExplSyntaxOn
\RenewDocumentCommand \num { O{} m o }
  {
    \mode_leave_vertical:
    \group_begin:
      \keys_set:nn { siunitx } {#1}
      % Is this a range or not?
      \IfNoValueTF {#3}
        {
          \tl_if_in:nnTF {#2} { x }
            { \numproduct {#2} }
            {
              \tl_if_in:nnTF {#2} { ; }
                { \numlist {#2} }
                {
                  \siunitx_number_format:nN {#2} \l__siunitx_tmp_tl
                  \siunitx_print_number:V \l__siunitx_tmp_tl 
                }
            }
        }
        { \numrange {#2} {#3} }
    \group_end:
  }
\ExplSyntaxOff

\begin{document}

\num{12345}

\num{1x2}

\num{1;2}

\num{1}[2]

\end{document}
```
As I said, there's then a performance hit for every occurrence of `\num`, and I feel questionable semantic meaning. (I did consider something like this in the v2 to v3 changes, but decided in the end to keep things separate as three is not a major gain.)

**Comment by dbitouze** [comment]
Thanks for the ready-to-use code!

>As I said, there's then a performance hit for every occurrence of \num, and I feel questionable semantic meaning. (I did consider something like this in the v2 to v3 changes, but decided in the end to keep things separate as three is not a major gain.)

Here is why I asked this feature request: just before I did it, I had been bitten by something like `\num{2x3}` since didn't remind I had to resort a different command for a product. And I had to look in the documentation to recall its exact name. To sum up: every user of `siunitx` will recall the names of the core commands `\num` and `\qty` but probably not all the other ones.
